funsor/minipyro.py

```diff
@@ -345,15 +345,18 @@ def step(self, *args, **kwargs):
 
 # TODO(eb8680) Replace this with funsor.Expectation.
 def Expectation(log_probs, costs, sum_vars, prod_vars):
-    probs = [p.exp() for p in log_probs]
     result = 0
     for cost in costs:
-        result += funsor.sum_product.sum_product(
-                sum_op=funsor.ops.add,
-                prod_op=funsor.ops.mul,
-                factors=probs + [cost],
-                plates=prod_vars,
-                eliminate=prod_vars | sum_vars)
+        log_prob = funsor.sum_product.sum_product(
```

Comment:
In `pyro.infer.util.Dice.compute_expectation`, we use `adjoint` to share intermediate computations for costs in the same ordinal.  We could do the same thing here.